streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -252,6 +288,111 @@ public void buildAndOptimizeTopology() {
         }
     }
 
+    private void maybePerformOptimizations(final Properties props) {
+
+        if (props != null && props.getProperty(StreamsConfig.TOPOLOGY_OPTIMIZATION).equals(StreamsConfig.OPTIMIZE)) {
+            LOG.debug("Optimizing the Kafka Streams graph for repartition nodes");
+            maybeOptimizeRepartitionOperations();
+        }
+    }
+
+    private void maybeOptimizeRepartitionOperations() {
+        StreamsGraphNode optimizedSingleRepartition = null;
+
+        for (final Map.Entry<StreamsGraphNode, Set<OptimizableRepartitionNode>> streamsGraphNodeSetEntry : keyChangingOperationsToOptimizableRepartitionNodes.entrySet()) {
+
+            StreamsGraphNode keyChangingNode = streamsGraphNodeSetEntry.getKey();
```

Comment:
final ;)

Reply:
ugh! I thought I got all of those, fixed